istioctl/cmd/describe.go

```diff
@@ -1186,3 +1197,94 @@ func containerReady(pod *v1.Pod, containerName string) (bool, error) {
 	}
 	return false, fmt.Errorf("no container %q in pod", containerName)
 }
+
+// describePeerAuthentication fetches all PeerAuthentication in workload and root namespace.
+// It lists the ones applied to the pod, and the current active mTLS mode.
+// When the client doesn't have access to root namespace, it will only show workload namespace Peerauthentications.
+func describePeerAuthentication(writer io.Writer, configClient istioclient.Interface, istioNamespace, workloadNamespace string, pod *v1.Pod, podsLabels k8s_labels.Set) error { // nolint: lll
+	workloadPAList, err := configClient.SecurityV1beta1().PeerAuthentications(workloadNamespace).List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch workload namespace PeerAuthentication: %v", err)
+	}
+
+	hasRootNamespaceAccess := true
+	rootPAList, err := configClient.SecurityV1beta1().PeerAuthentications(istioNamespace).List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		hasRootNamespaceAccess = false
+		fmt.Fprintf(writer, "Cannot analyze active mTLS mode since failed to fetch root namespace PeerAuthentication: %v:\n", err)
+	}
+
+	allPAs := append(rootPAList.Items, workloadPAList.Items...)
+	if len(allPAs) == 0 {
+		return nil
+	}
+
+	var cfgs []*config.Config
+	for _, pa := range allPAs {
+		pa := pa
+		cfg := crdclient.TranslateObject(&pa, config.GroupVersionKind(pa.GroupVersionKind()), "")
+		cfgs = append(cfgs, &cfg)
+	}
+
+	matchedPA := findMatchedConfigs(podsLabels, cfgs)
+	printConfigs(writer, matchedPA)
+
+	if hasRootNamespaceAccess {
+		activePA := authnv1beta1.ComposePeerAuthentication(istioNamespace, matchedPA)
+		printPeerAuthentication(writer, activePA)
+	}
+
+	return nil
+}
+
+// Workloader is used for matching all configs
+type Workloader interface {
+	GetSelector() *typev1beta1.WorkloadSelector
+}
+
+// findMatchedConfigs should filter out unrelated configs that are not matched given podsLabels.
+// Configs passed into this method should only contains workload's namespaces configs
+// and rootNamespaces configs, caller should be responsible for controlling configs passed
+// in.
+func findMatchedConfigs(podsLabels k8s_labels.Set, configs []*config.Config) []*config.Config {
```

Comment:
Could you clarify what does the `Configs passed into this method should only contains workload's namespaces configs and rootNamespaces configs` mean? What's the negative cases?

Reply:
Added more comments, basically the idea is when there is no workload selector, it will treat as qualified, meaning it is either workload namespace or root namespace.